### ISSUE TITLE
[FR] Record model signatures and input examples for TensorFlow & PySpark ML autologging integrations

### ISSUE THREAD

**Issue by dbczumar** [mlflow/mlflow#4828]
## MLflow Roadmap Item

This is an MLflow Roadmap item that has been prioritized by the MLflow maintainers. We're seeking help with the implementation of roadmap items tagged with the `help wanted` label.

For requirements clarifications and implementation questions, or to request a PR review, please tag @BenWilson2 in your communications related to this issue.

## Proposal Summary

Include model signature and input example information with MLflow Models that are logged during autologging. This functionality is currently only present in a few autologging integrations: [mlflow.xgboost.autolog()](https://github.com/mlflow/mlflow/blob/94891da1c272038ac506f5d091c8c6b19e009b2d/mlflow/xgboost.py#L662-L664), [mlflow.lightgbm.autolog()](https://github.com/mlflow/mlflow/blob/94891da1c272038ac506f5d091c8c6b19e009b2d/mlflow/lightgbm.py#L553-L555), and [mlflow.sklearn.autolog()](https://github.com/mlflow/mlflow/blob/94891da1c272038ac506f5d091c8c6b19e009b2d/mlflow/sklearn/__init__.py#L1227-L1229), but there are many other integrations listed here - https://mlflow.org/docs/latest/tracking.html#automatic-logging - that do not have this functionality. Most prominently, we should add support for TensorFlow and PySpark ML.

## Motivation
- What is the use case for this feature? Model signatures and input examples make it easier to incorporate ML models into inference workflows.
- Why is this use case valuable to support for MLflow users in general? ^
- Why is this use case valuable to support for your project(s) or organization? ^
- Why is it currently difficult to achieve this use case? Users must currently compute signatures / input examples manually and invoke `mlflow.*.log_model()` to record this information with a persisted MLflow Model. This complicates the autologging experience.

### What component(s), interfaces, languages, and integrations does this feature affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interfaces
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Languages 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by bipinKrishnan** [comment]
Hey @dbczumar, I'm new here, can I work on this issue? If possible can you guide me on what needs to be done to possibly fix this issue?

**Comment by dbczumar** [comment]
Hi @bipinKrishnan . Absolutely! I think a good place to start would be working on adding model signature and input example information to a single autologging integration: `mlflow.tensorflow.autolog()`. TensorFlow autologging supports two model types: `tf.Estimator` models and `tf.keras` models. To begin with, let's focus on `tf.keras` models.

The goal is to write a PR that updates the autologging routine for `tf.keras` (https://github.com/mlflow/mlflow/blob/e25910bd5bb750222dd9c316ad132b99b66c707e/mlflow/tensorflow.py#L810-L812 / https://github.com/mlflow/mlflow/blob/e25910bd5bb750222dd9c316ad132b99b66c707e/mlflow/tensorflow.py#L851-L853) to capture a slice of the model input and infer an MLflow model signature from the input using the `mlflow.models.infer_signature()` method (https://www.tensorflow.org/api_docs/python/tf/keras/Model#fit). The input and inferred signature should then be passed to the `log_model()` API call used to persist the model. You can find a similar example for scikit-learn models here: https://github.com/mlflow/mlflow/blob/e25910bd5bb750222dd9c316ad132b99b66c707e/mlflow/sklearn/__init__.py#L1219-L1229, which was introduced via https://github.com/mlflow/mlflow/pull/3326.

As seen in the `tf.keras.fit()` method (https://www.tensorflow.org/api_docs/python/tf/keras/Model#fit), there are several accepted input types for the `x` and `y` dataset parameters. We should try to introduce logic to extract numpy arrays or dictionaries of { name -> numpy.ndarray} for as many of these types as possible, since `mlflow.models.infer_signature()` only accepts certain input types (see https://mlflow.org/docs/latest/python_api/mlflow.models.html#mlflow.models.infer_signature). If there's a data type we can't   infer a signature from, we can just print a warning.

**Comment by bipinKrishnan** [comment]
Thanks @dbczumar for the detailed explanation, I will take a look into it.

**Comment by bipinKrishnan** [comment]
@dbczumar @BenWilson2 After doing some experimentation, here are some of my thoughts.
Since the inputs of the model during training are recorded inside the ```*args``` and ```**kwargs``` of ```train``` method below, we can just filter out the input examples from them and use it to infer the model signature and then log a slice of the input and model signature to mlflow. Please feel free to correct me if I'm doing it the wrong way.

https://github.com/mlflow/mlflow/blob/e25910bd5bb750222dd9c316ad132b99b66c707e/mlflow/tensorflow.py#L971-L984

**Comment by dbczumar** [comment]
@bipinKrishnan Your described approach sounds great! Feel free to try implementing it. Please let me know if you have any questions!

**Comment by bipinKrishnan** [comment]
@dbczumar I've logged the model signature using ```save_model``` function inside ```export_saved_model``` function. My question is that, is it okay to do the logging this way or should we use ```mlflow.tensorflow.log_model``` to log model signatures?

https://github.com/mlflow/mlflow/blob/e25910bd5bb750222dd9c316ad132b99b66c707e/mlflow/tensorflow.py#L1020-L1025

**Comment by dbczumar** [comment]
@bipinKrishnan Good question! The `mlflow.tensorflow.log_model()` function should be used to log the model with a signature.

**Comment by bipinKrishnan** [comment]
> @dbczumar I've logged the model signature using `save_model` function inside `export_saved_model` function and its working fine. My question is that, is it okay to do the logging this way or should we use `mlflow.tensorflow.log_model` to log model signatures?
> 
> https://github.com/mlflow/mlflow/blob/e25910bd5bb750222dd9c316ad132b99b66c707e/mlflow/tensorflow.py#L1020-L1025

@dbczumar Just to clarify, so, should I overwrite ```save_model``` with ```log_model``` ?

**Comment by BenWilson2** [comment]
The current implementation in `export_saved_model` is effectively doing what a log_model() call would do, with the exception of registration as a pyfunc (instead of exclusively saved as the flavor). Adding signature inference with a fallback of `None` and example input samples to support using `log_model()` seems to be the best approach to work on here. 
Let us know if you'd like a set of eyes on any ideations you're doing for signature inference or if you'd like someone to bounce an idea off of.

**Comment by bipinKrishnan** [comment]
@BenWilson2 I'm currently writing the tests, but the ```create_tf_estimator_model``` does not allow us to access the model or the training data used. But I need to access these to compare the model signature and input example logged by ```mlflow.tensorflow.autolog```.
Is there any work around for this, should I go ahead and modify ```create_tf_estimator_model``` and make it return the required items?

**Comment by BenWilson2** [comment]
If you look at the tests below that utilize `create_tf_estimator_model()`, you'll notice that they're all using that function as a script execution, with no consumption of the `None` return. 
You could refactor that by creating a dataclass (or namedtuple) that holds the values that you need for your signature and example tests for currying to your test modules, rename the script execution function, e.g., `def _create_tf_estimator_model() -> <your data class>:`, then create a handler function `def create_tf_estimator_model() -> None:` that maintains parity with the rest of the tests. For your signature and example tests you could then call the dataclass-returning private function to consume the return values for assertions. 
Hope that helps!

**Comment by bali0019** [comment]
@BenWilson2 After my chat with @dbczumar, I will be taking on pyspark.ml autologging integration...fyi.

**Comment by dbczumar** [comment]
Closing this out now that https://github.com/mlflow/mlflow/pull/5719, https://github.com/mlflow/mlflow/pull/5510, and https://github.com/mlflow/mlflow/pull/5461 have been merged, which introduce signature / input example autologging for TensorFlow and PySpark ML. Thanks a bunch, @bali0019 !
